### PR TITLE
Introduce LDAP Group Provider

### PR DESCRIPTION
**Pull request by rimolive**
This PR supersedes #8335

### CODE REVIEW

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java [hunk 1]
```diff
@@ -106,11 +111,44 @@ public boolean isGroupMember(String searchBase, String groupSearch, String conte
...
+        }
+    }
+
     private static CloseableSearchResults searchContext(String searchBase, String searchFilter, CloseableContext context)
             throws NamingException
     {
         SearchControls searchControls = new SearchControls();
         searchControls.setSearchScope(SearchControls.SUBTREE_SCOPE);
+        searchControls.setReturningAttributes(new String[]
+        {
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java:149]: nit: We can inline this.

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticator.java [hunk 2]
```diff
@@ -162,22 +161,6 @@ private Principal authenticateWithBindDistinguishedName(Credential credential)
...
 
-    /**
-     * Returns {@code true} when parameter contains a character that has a special meaning in
-     * LDAP search or bind name (DN).
-     * <p>
-     * Based on <a href="https://www.owasp.org/index.php/Preventing_LDAP_Injection_in_Java">Preventing_LDAP_Injection_in_Java</a> and
-     * {@link javax.naming.ldap.Rdn#escapeValue(Object) escapeValue} method.
-     */
-    @VisibleForTesting
-    static boolean containsSpecialCharacters(String user)
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticator.java:173]: Can we move this to a separate commit ? So that it would be easy to review.

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java [hunk 3]
```diff
@@ -49,16 +54,16 @@
 import static javax.naming.Context.SECURITY_CREDENTIALS;
 import static javax.naming.Context.SECURITY_PRINCIPAL;
 
-public class JdkLdapAuthenticatorClient
-        implements LdapAuthenticatorClient
+public class JdkLdapClient
+        implements LdapClient
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java:58]: Rename can be handled in a new commit ?

**Review by Praveen2112 — COMMENTED** [review]
Just skimmed it.

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java [hunk 4]
```diff
@@ -106,11 +111,44 @@ public boolean isGroupMember(String searchBase, String groupSearch, String conte
         }
     }
 
+    @Override
+    public Set<String> lookupUserGroups(String searchBase, String searchFilter, String contextUserDistinguishedName, String contextPassword)
+            throws NamingException
+    {
+        try (CloseableContext context = createUserDirContext(contextUserDistinguishedName, contextPassword); CloseableSearchResults search = searchContext(searchBase, searchFilter, context)) {
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java:118]:
```suggestion
        try (CloseableContext context = createUserDirContext(contextUserDistinguishedName, contextPassword); 
                CloseableSearchResults search = searchContext(searchBase, searchFilter, context)) {
```

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java [hunk 5]
```diff
@@ -106,11 +111,44 @@ public boolean isGroupMember(String searchBase, String groupSearch, String conte
...
+            ImmutableSet.Builder<String> groupNames = ImmutableSet.builder();
+            if (search.hasMore()) {
+                Attributes attributes = search.next().getAttributes();
+                Attribute memberOfAttribute = attributes.get("memberof");
+                if (memberOfAttribute == null) {
+                    throw new NamingException("No memberOf attribute found... The ldap group provider requires the memberOf overlay to be enabled.");
+                }
+                else {
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java:126]: We can remove this else

File: plugin/trino-password-authenticators/src/test/java/io/trino/plugin/password/ldap/TestLdapUtil.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestLdapUtil
+{
+    @Test
+    public void testContainsSpecialCharacters()
+    {
+        assertThat(LdapUtil.containsSpecialCharacters("The quick brown fox jumped over the lazy dogs"))
+                .as("English pangram")
+                .isEqualTo(false);
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/test/java/io/trino/plugin/password/ldap/TestLdapUtil.java:27]: Can we use `isTrue/isFalse`

File: plugin/trino-password-authenticators/src/test/java/io/trino/plugin/password/ldap/TestLdapGroupProvider.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+        LdapGroupProvider ldapGroupProvider = new LdapGroupProvider(
+                client,
+                new LdapConfig()
+                        .setUserBaseDistinguishedName("base-dn")
+                        .setUserBindSearchPatterns("${USER}@example.com")
+                        .setBindDistingushedName("server")
+                        .setBindPassword("server-pass"));
+
+        assertEquals(ldapGroupProvider.getGroups("alice"), ImmutableSet.of("group_a", "group_b"));
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/test/java/io/trino/plugin/password/ldap/TestLdapGroupProvider.java:38]: We can use `assertThat(..).containsExactly(..)`

File: plugin/trino-password-authenticators/src/test/java/io/trino/plugin/password/ldap/TestLdapClient.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+
+public class TestLdapClient
+        implements LdapClient
+{
+    private static final String BASE_DN = "base-dn";
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/test/java/io/trino/plugin/password/ldap/TestLdapClient.java:30]: Can we use a builder based model for this ? We can add credentials and group memeber as we require and then finally build the TestLdapClient.

Comment (Praveen2112) [reply plugin/trino-password-authenticators/src/test/java/io/trino/plugin/password/ldap/TestLdapClient.java:30]: But it involves changing the existing tests

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticatorFactory.java [hunk 9]
```diff
@@ -29,7 +29,7 @@
     @Override
     public String getName()
     {
-        return "ldap";
+        return "ldap-trino";
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticatorFactory.java:32]: Do we need this change ?

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    private final Optional<String> userBaseDistinguishedName;
+    private final Optional<String> bindDistinguishedName;
+    private final Optional<String> bindPassword;
+
+    @Inject
+    public LdapGroupProvider(LdapClient client, LdapConfig ldapConfig)
+    {
+        this.client = requireNonNull(client, "client is null");
+
+        this.userBindSearchPatterns = ldapConfig.getUserBindSearchPatterns();
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:48]: requireNonNull checks would be helpful. And a `Immutable#copy`

Comment (Praveen2112) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:48]: non null checks for `ldapConfig`

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        this.userBaseDistinguishedName = Optional.ofNullable(ldapConfig.getUserBaseDistinguishedName());
+        this.bindDistinguishedName = Optional.ofNullable(ldapConfig.getBindDistingushedName());
+        this.bindPassword = Optional.ofNullable(ldapConfig.getBindPassword());
+
+        checkArgument(userBaseDistinguishedName.isPresent(), "Base distinguished name (DN) for user must be provided");
+        checkArgument(bindDistinguishedName.isPresent() == bindPassword.isPresent(), "Both bind distinguished name and bind password must be provided");
+        checkArgument(!userBindSearchPatterns.isEmpty(), "User bind search pattern must be provided");
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:55]: Can we have it as a part of config assertion ?

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    @Override
+    public Set<String> getGroups(String user)
+    {
+        if (LdapUtil.containsSpecialCharacters(user)) {
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:61]: static import

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+            throw new AccessDeniedException("Username contains a special LDAP character");
+        }
+        for (String userBindSearchPattern : userBindSearchPatterns) {
+            String userDistinguishedName = LdapUtil.replaceUser(userBindSearchPattern, user);
+            String searchBase = userBaseDistinguishedName.orElseThrow();
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:66]: What if `userBaseDistinguishedName` is not provided ? If we are throwing an exception, can we move check to constructor ?

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapUtil.java [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.password.ldap;
+
+import com.google.common.base.CharMatcher;
+
+public final class LdapUtil
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapUtil.java:18]: Moving to Util clas can be in a separate commit

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java [hunk 15]
```diff
@@ -106,11 +108,30 @@ public boolean isGroupMember(String searchBase, String groupSearch, String conte
...
+    {
+        try (CloseableContext context = createUserDirContext(contextUserDistinguishedName, contextPassword);
+                CloseableSearchResults search = searchContext(searchBase, searchFilter, context)) {
+            ImmutableSet.Builder<String> groupNames = ImmutableSet.builder();
+            if (search.hasMore()) {
+                Attributes attributes = search.next().getAttributes();
+                Attribute memberOfAttribute = attributes.get("memberof");
+                if (memberOfAttribute == null) {
+                    throw new NamingException("No memberOf attribute found... The ldap group provider requires the memberOf overlay to be enabled.");
+                }
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java:123]: Where are we adding the group names here ?

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java [hunk 16]
```diff
@@ -106,11 +108,30 @@ public boolean isGroupMember(String searchBase, String groupSearch, String conte
...
+            return groupNames.build();
+        }
+    }
+
     private static CloseableSearchResults searchContext(String searchBase, String searchFilter, CloseableContext context)
             throws NamingException
     {
         SearchControls searchControls = new SearchControls();
         searchControls.setSearchScope(SearchControls.SUBTREE_SCOPE);
+        searchControls.setReturningAttributes(new String[] {"memberOf"});
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java:134]: Will it affect the `lookupUserDistinguishedNames` ?

Comment (Praveen2112) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java:134]: From the JavaDocs it looks like 
```
Specifies the attributes that will be returned as part of the search.
null indicates that all attributes will be returned. An empty array indicates no attributes are returned.
Params:
attrs – An array of attribute ids identifying the attributes that will be returned. Can be null.
See Also:
getReturningAttributes
```

Is there is any problem on returning all the attributes ?

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java (hunk 15 above)
Comment (Praveen2112) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java:123]: So we don't support groups from LDAP server doesn't provide the `memberof` ? In that case can we fetch the groups based on DN of the user ?

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticatorFactory.java (hunk 9 above)
Comment (rimolive) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticatorFactory.java:32]: It was requested earlier as it conflicts with the Presto LDAP Group Provider.

Comment (Praveen2112) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticatorFactory.java:32]: I think it can be `ldap` so as to be consistent with other password-authenticator.

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticatorFactory.java [hunk 17]
```diff
@@ -39,7 +39,7 @@ public PasswordAuthenticator create(Map<String, String> config)
                 binder -> {
                     configBinder(binder).bindConfig(LdapConfig.class);
                     binder.bind(LdapAuthenticator.class).in(Scopes.SINGLETON);
-                    binder.bind(LdapAuthenticatorClient.class).to(JdkLdapAuthenticatorClient.class).in(Scopes.SINGLETON);
+                    binder.bind(LdapClient.class).to(JdkLdapClient.class).in(Scopes.SINGLETON);
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticatorFactory.java:42]: I think it has to be a part of this commit. https://github.com/trinodb/trino/pull/10116/commits/d3f7eff6991c79e94be1db675aa5fe963374ef8c

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapClient.java [hunk 18]
```diff
@@ -17,7 +17,7 @@
 
 import java.util.Set;
 
-public interface LdapAuthenticatorClient
+public interface LdapClient
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapClient.java:20]: I think it has to be part of this commit - https://github.com/trinodb/trino/pull/10116/commits/d3f7eff6991c79e94be1db675aa5fe963374ef8c

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java (hunk 10 above)
Comment (Praveen2112) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:48]: ping

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java (hunk 11 above)
Comment (Praveen2112) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:55]: ping

File: plugin/trino-password-authenticators/src/test/java/io/trino/plugin/password/ldap/TestLdapClient.java [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+import io.trino.plugin.password.Credential;
+
+import javax.naming.NamingException;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+
+public class TestLdapClient
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/test/java/io/trino/plugin/password/ldap/TestLdapClient.java:27]: We don't need this. Can we use the dockerized open ldap server to add users and groups ?

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+            try {
+                return client.lookupUserGroups(searchBase, userDistinguishedName, bindDistinguishedName.orElseThrow(), bindPassword.orElseThrow());
+            }
+            catch (NamingException e) {
+                log.debug(e, "Authentication failed for user [%s], %s", user, e.getMessage());
+                throw new RuntimeException("Authentication error");
```
[the remark this answers is not in the document]
Comment (Praveen2112) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:72]: I think it is better to throw an exception if the user doesn't exist.

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java [hunk 21]
```diff
@@ -106,11 +108,30 @@ public boolean isGroupMember(String searchBase, String groupSearch, String conte
         }
     }
 
+    @Override
+    public Set<String> lookupUserGroups(String searchBase, String searchFilter, String contextUserDistinguishedName, String contextPassword)
+            throws NamingException
+    {
+        try (CloseableContext context = createUserDirContext(contextUserDistinguishedName, contextPassword);
+                CloseableSearchResults search = searchContext(searchBase, searchFilter, context)) {
+            ImmutableSet.Builder<String> groupNames = ImmutableSet.builder();
```
[the remark this answers is not in the document]
Comment (Praveen2112) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java:117]: ping

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java [hunk 22]
```diff
@@ -0,0 +1,248 @@
...
+            if (search.hasMore()) {
+                Attributes attributes = search.next().getAttributes();
+                Attribute memberOfAttribute = attributes.get("memberof");
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java:120]: Can we make this configurable ?

Comment (Praveen2112) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java:120]: What if the LDAP Server doesn't have this `memberOf` overlay ??
Can we have a fallback mechanism where we could search for a set of groups and then search if the users exists in that group ?

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (Praveen2112) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java:123]: It is better if we handle these condition too.

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java (hunk 13 above)
Comment (rimolive) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:66]: Didn't we already check in the constructor? Anyways, I'll just change it to `get()`

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapUtil.java (hunk 14 above)
Comment (rimolive) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapUtil.java:18]: I still don't get that rule. `LdapUtil` was introduced as well as other classes in the first commit. Why it requires a separate commit?

**Comment by rimolive** [comment]
@Praveen2112 Just FYI I'm setting this PR to WIP to rebase latest changes and refactor the code based on that rebase.

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapCommon.java [hunk 23]
```diff
@@ -13,26 +13,52 @@
...
 import javax.naming.NamingEnumeration;
 import javax.naming.NamingException;
 
 import java.util.Set;
 
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.lang.String.format;
 import static java.util.Objects.requireNonNull;
 
-public class LdapAuthenticatorClient
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapCommon.java:28]: Do we need this renaming ?

Comment (Praveen2112) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapCommon.java:28]: Can we have a client extensively for group provider ?

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java [hunk 24]
```diff
@@ -0,0 +1,148 @@
...
+    @VisibleForTesting
+    void invalidateCache()
+    {
+        groupCache.invalidateAll();
+    }
+
+    @Override
+    public Set<String> getGroups(String user)
+    {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:88]: Did we see any missing class error ?

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java [hunk 25]
```diff
@@ -0,0 +1,148 @@
...
+    private final Optional<String> userBaseDistinguishedName;
+    private final Optional<String> groupBaseDistinguishedName;
+    private final Optional<String> bindDistinguishedName;
+    private final Optional<String> bindPassword;
+    private final boolean allowUserNotExist;
+
+    private final LoadingCache<String, Set<String>> groupCache;
+
+    @Inject
+    public LdapGroupProvider(LdapCommon client, LdapClientConfig ldapConfig)
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:60]: We can use `LdapClient` instead of `LdapCommon`

**Comment by Praveen2112** [comment]
The jar files are not required. Can we remove them ?

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java (hunk 24 above)
Comment (rimolive) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:97]: As I told you on Slack, I'm getting errors while running the tests. I'm not 100% this will get a missing class error, but I'm looking to fix the test locally.

**Comment by rimolive** [comment]
@Praveen2112 I'm not sure why, but this PR is removing the jars. So other commit added them by mistake.

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java (hunk 24 above)
Comment (rimolive) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:96]: Update on this: I found the problem and I can run the tests locally. I'll be working on fixing the tests today.

**Comment by rimolive** [comment]
@woowahan-jaehoon It would be good if you can review the code too.

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/ldap/LdapClientConfig.java [hunk 26]
```diff
@@ -25,19 +27,30 @@
...
 public class LdapClientConfig
 {
     private String ldapUrl;
     private boolean allowInsecure;
+    private List<String> userBindSearchPatterns = ImmutableList.of();
+    private String groupAuthorizationSearchPattern;
+    private String userBaseDistinguishedName;
+    private String groupBaseDistinguishedName;
+    private String bindDistinguishedName;
+    private String bindPassword;
```
[the remark this answers is not in the document]
Comment (Praveen2112) [reply lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/ldap/LdapClientConfig.java:46]: Any reason for not having a dedicated config class ?

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java (hunk 24 above)
Comment (Praveen2112) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:95]: Do we still see this missing class error ?

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/ldap/LdapClientConfig.java (hunk 26 above)
[the remark this answers is not in the document]
Comment (Praveen2112) [reply lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/ldap/LdapClientConfig.java:46]: Can we remove these changes as we have a dedicated config class for the same ?

**Comment by Praveen2112** [comment]
Can you remove the jars from the commit ?

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java [hunk 27]
```diff
@@ -0,0 +1,211 @@
...
+        this.client = requireNonNull(client, "client is null");
+        requireNonNull(ldapConfig, "ldapConfig is null");
+
+        this.userBaseDistinguishedName = Optional.ofNullable(ldapConfig.getUserBaseDistinguishedName());
+        this.groupBaseDistinguishedName = Optional.ofNullable(ldapConfig.getGroupBaseDistinguishedName());
+        this.bindDistinguishedName = Optional.ofNullable(ldapConfig.getBindDistinguishedName());
+        this.bindPassword = Optional.ofNullable(ldapConfig.getBindPassword());
+        this.allowUserNotExist = ldapConfig.isAllowUserNotExist();
+
+        checkArgument(bindDistinguishedName.isPresent() == bindPassword.isPresent(), "Both bind distinguished name and bind password must be provided");
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:77]: Can we make them non-Optional and maybe annotate with `@NotNull` for the getter method ?

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java [hunk 28]
```diff
@@ -0,0 +1,211 @@
...
+        checkArgument(groupBaseDistinguishedName.isPresent(), "Base distinguished name (DN) for group must be provided");
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:78]: ditto

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java [hunk 29]
```diff
@@ -0,0 +1,219 @@
...
+                        this.bindDistinguishedName,
+                        this.bindPassword);
+                String groupFilter = replaceUser(this.groupSearchFilter, userDistinguishedName);
+                log.debug("Querying LDAP from %s using filter %s", this.groupBaseDistinguishedName, groupFilter);
+                Set<String> queryResult = lookupUserDistinguishedNames(
+                        this.groupBaseDistinguishedName, groupFilter,
+                        bindDistinguishedName, bindPassword);
+                groupSet = queryResult.stream().map(source -> {
+                    try {
+                        log.debug("extracting group from %s", source);
```
Comment (Praveen2112) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:135]: Instead of matching pattern can we fetch the actual attributes ?

Comment (rimolive) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:127]: What attributes are you talking about? Depending on what they are, we can add more debug logs if needed.

**Comment by bitsondatadev** [comment]
:wave: @rimolive - this PR has become inactive. If you're still interested in working on it, please let us know, and we can try to get reviewers to help with that.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.

**Comment by woowahan-jaehoon** [comment]
> @woowahan-jaehoon It would be good if you can review the code too.

I left early this year and I changed job. So I cannot review this PR right now.

I asked to test this PR to teammate who in my before company already but I didn't get that result yet. (cc: @iamhongri)

Maybe I'll return before job, I'll use this usefully. Thank you.

**Comment by colebow** [comment]
Closing this one out due to inactivity, but please reopen if you would like to pick this back up.

**Comment by dprophet** [comment]
@bitsondatadev What happened with this PR? We also have an LDAP system our internal Bloomberg build of Trino. Perhaps we should make more efforts to get this released to the community.

**Comment by Praveen2112** [comment]
@dprophet The PR was inactive so it was closed. Maybe we could re-open it if someone is ready to work on it.

**Comment by cs-ps** [comment]
This is functionality we'd use. +1 to reopening if there's availability.

**Comment by bitsondatadev** [comment]
It looks like the PR author @rimolive is no longer working on this and hasn't responded so far. If Bloomberg or anyone else wants to open up a PR that contains this work please do and make sure to reference this PR when opening it so that anyone who comes back to this PR will be directed to the new active PR. This PR will only be reopened if @rimolive wants to pick the work back up. I wouldn't wait though since he hasn't responded since we asked in November of last year.

Happy to provide guidance if anyone needs help.

**Comment by accorvin** [comment]
Hi all, I work with @rimolive and can chime in on his behalf. We have been running a development build of this change set in our production Trino deployment for several months, so I can confirm it's stable.

@Praveen2112 I know that @rimolive felt that too much functionality was being requested as part of this initial PR, and that was a hinderance to him being able to get it done (see [this comment](https://github.com/trinodb/trino/pull/10116#discussion_r953547224) for an example).

I can ask @rimolive to reopen this PR if the community desires, although our priorities have shifted and we therefore don't have time to extend/update the existing functionality beyond the current change set here.

**Comment by bitsondatadev** [comment]
I'll leave this up to @Praveen2112.

It sounds like this would only be possible if the PR is mostly merged as-is. Praveen if you feel like this PR is pretty close then maybe we can ask @rimolive to do a few small changes and merge it.

If, however, you feel strongly that there's still much that needs to be addressed, perhaps we should let someone else in the community take this one over and run with it.

**Comment by dprophet** [comment]
We have also been using a similar fork of this original plugin internally for quite some time. We are running into needs to formalize this plugin to feed connector specific io.trino.spi.connector.ConnectorAccessControl. Maintaining a plugin outside of the main Trino code base is darn expensive.

**Comment by bitsondatadev** [comment]
After speaking with @Praveen2112, it seems like there are some changes needed before this is ready to be merged. So merging as-is will not be an option. Anyone who is interested in picking up this work is highly encouraged to do so! If you need help, please reach out to me or on the `#dev` channel on the [Trino Slack](https://trino.io/slack.html).
